dhis-2/dhis-api/src/main/java/org/hisp/dhis/eventdatavalue/EventDataValueService.java

```diff
@@ -0,0 +1,100 @@
+package org.hisp.dhis.eventdatavalue;
+/*
+ * Copyright (c) 2004-2018, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.program.ProgramStageInstance;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author David Katuscak
+ */
+public interface EventDataValueService
```

Comment:
Very interesting question. The answer is not easy. Probably I "got tunneled" and went this way because of the logic around like audit, files handling, validation, etc.

But I see your point. Probably I could dive back into the code and rewrite it in a way that it is updated directly via event while all the validation, audit, etc. is still done.

Reply:
If they work with ProgramStageInstance object, that means they are working with Event. 

Also should they be dictating how EventService should be architected?

By the way I have never seen a case where users are importing a filled PDF containing an event data. I don't think we have a UI to handle that... so I guess you can kind of remove that class and go ahead with the "normal" service behavior of EventService

Reply:
The problem is, that when it is coming from DTO (dxf2) layer, so working with Event, I do validation logic on DTO layer. 

But when it is coming from DAO layer, so working with ProgramStageInstance, then I have to validate on DAO layer. Therefore, there are separate method for it. I think, and I prefer, that business logic (validation, access rights checking, etc.) should be done on DTO layer, however, those 2 named classes don't work with DTO. Therefore, I implemented it as it is. The ideal solution is to remove classes if they are not in use, respectively rewrite them to use DTO and not access DAO directly.

Regarding PDF and SMS -> I just rewrote previous logic into a new one. So the data values manipulation was there from before. Should be removed if not in use or impossible to use it with current UI.